src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenInParametersTests.cs

```diff
@@ -4386,5 +4386,193 @@ .locals init (Test V_0)
 }
 ");
         }
+
+        [Fact, WorkItem(66135, "https://github.com/dotnet/roslyn/issues/66135")]
+        public void ConstrainedCallOnInParameter()
+        {
+            var source = @"
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+
+public class C
+{
+    public static void Main()
+    {
+        S value = new();
+        ref readonly S valueRef = ref value;
+        Console.Write(valueRef);
+        M(in valueRef);
+        Console.Write(valueRef);
+    }
+    public static void M(in S value)
+    {
+        foreach (var x in value) { }
+    }
+}
+
+public struct S : IEnumerable<int>
+{
+    int a;
+    public readonly override string ToString() => a.ToString();
+    private IEnumerator<int> GetEnumerator() => Enumerable.Range(0, ++a).GetEnumerator();
+    IEnumerator<int> IEnumerable<int>.GetEnumerator() => GetEnumerator();
+    IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
+}";
+            var verifier = CompileAndVerify(source, expectedOutput: "00");
+            // Note: we use a temp instead of directly doing a constrained call on `in` parameter
```

Comment:
Do we use a temp even when the target method is marked as `readonly`?

Reply:
It's not possible to have exactly this scenario with the only modification being an additional `readonly` because this scenario uses an interface and `readonly` is only allowed in `struct` types (not in interfaces).
That said, it should be possible to emit a pattern-based `GetEnumerator` call directly on the struct (ie. `foreach` without `IEnumerable` interface). I'll add a test for that.

Reply:
FYI, amended my earlier comment